### PR TITLE
Refactor telemetry manager and add more tests

### PR DESCRIPTION
**Pull request by Evangelink**

### CODE REVIEW

File: src/Integration/Telemetry/TelemetryManager.cs [hunk 1]
```diff
@@ -67,92 +62,86 @@ public bool IsAnonymousDataShared
...
-            string path = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "msenv.dll");
-            visualStudioVersion = File.Exists(path)
-                ? FileVersionInfo.GetVersionInfo(path).ProductVersion
-                : string.Empty;
+            this.knownUIContexts = knownUIContexts;
 
             if (this.telemetryRepository.Data.InstallationDate == DateTime.MinValue)
             {
-                this.telemetryRepository.Data.InstallationDate = DateTime.Now;
+                this.telemetryRepository.Data.InstallationDate = DateTime.Now; // TODO: Use some mockable clock
```
Comment (SonarTech) [inline src/Integration/Telemetry/TelemetryManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: src/Integration/Telemetry/TelemetryManager.cs (hunk 1 above)
Comment (SonarTech) [inline src/Integration/Telemetry/TelemetryManager.cs]: INFO Complete the task associated to this 'TODO' comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1135)

File: src/Integration/Telemetry/TelemetryManager.cs [hunk 2]
```diff
@@ -67,92 +62,86 @@ public bool IsAnonymousDataShared
...
-            return new TelemetryPayload
-            {
-                SonarLintProduct = "SonarLint Visual Studio",
-                SonarLintVersion = SonarLintVersion,
-                VisualStudioVersion = visualStudioVersion,
-                NumberOfDaysSinceInstallation = numberOfDaysSinceInstallation,
-                NumberOfDaysOfUse = this.telemetryRepository.Data.NumberOfDaysOfUse,
-                IsUsingConnectedMode = this.solutionBindingTracker.IsActiveSolutionBound
-            };
+            return FileVersionInfo.GetVersionInfo(Assembly.GetExecutingAssembly().Location).FileVersion;
```
Comment (valhristov) [inline src/Integration/Telemetry/TelemetryManager.cs]: Apparently this works, but why not use typeof(SonarLintSomething).Assembly.Location? [MSDN says](https://msdn.microsoft.com/en-us/library/system.reflection.assembly.getexecutingassembly(v=vs.110).aspx) you should not call this method when you know what is the executing assembly...

File: src/Integration/Helpers/ITimer.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Timers;
+
+namespace SonarLint.VisualStudio.Integration
+{
+    public interface ITimer
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply src/Integration/Helpers/ITimer.cs]: Why?

**Comment by Evangelink** [comment]
couldn't rebase properly so I have created a new branch and cherry picked the interesting commit. See #300

**Pull request closed without merging**